blocks/initialization.py

```diff
@@ -147,21 +148,39 @@ def __init__(self, mult=1):
         self.mult = mult
 
     def generate(self, rng, shape):
-        if len(shape) != 2:
-            raise ValueError
-        rows, cols = shape
-        return self.mult * numpy.eye(rows, cols, dtype=theano.config.floatX)
+        if len(shape) == 1:
+            eye_matrix = numpy.eye(1, shape[0], dtype=theano.config.floatX)
+            return self.mult * eye_matrix[0]
+        elif len(shape) == 2:
+            rows, cols = shape
+            eye_matrix = numpy.eye(rows, cols, dtype=theano.config.floatX)
+            return self.mult * eye_matrix
+        else:
+            raise ValueError("Identity initialization only works for 1D and"
+                             " 2D variables")
 
 
 class Orthogonal(NdarrayInitialization):
-    """Initialize a random orthogonal matrix.
+    r"""Initialize a random, approximately orthogonal matrix.
+
+    Only works for 1D and 2D variables.
 
-    Only works for 2D arrays.
+    When used to initialize a 1D parameter vector the elements will
+    be drawn i.i.d. from a normal distribution.
+
+    For 2D matrices we initialalize such that
+    :math:`M M^T \approx I` and :math:`M^T M \approx I` where
+    :math:`I` is an identity matrix of appropriate size.
 
     """
     def generate(self, rng, shape):
-        if len(shape) != 2:
-            raise ValueError
+        if len(shape) not in (1, 2):
+            raise ValueError("Orthogonal initialization only works for 1D and"
+                             " 2D variables")
+
+        if len(shape) == 1:
+            # Special case for 1D variables
+            return rng.randn(*shape).astype(theano.config.floatX)
```

Comment:
If we stick to the exact correspondence between the name and action taken, then only square matrices can be orthogonal.

But I am not sure that you are right, @bartvm. If the initialized object is a `1 x n` matrix, then indeed the expected operation with it is the dot product, and the logical extension of the orthogonality concept would be to make it have norm 1. But if it is a Theano vector, then most likely nobody's going to compute dot products with it. Element-wise multiplication is a more typical usage of such a tensor. And if we ask ourselves a question how a vector should be initialized so that element-wise multiplication with this vector would not change its norm, one of the possible answers is to make his elements i.i.d with zero mean and unit variance. In fact this would generalize to any tensor with number of dimensions other than two.

Apart from that, my main practical consideration: we should not make it possible that switching from `SimpleRecurrent(weights_init=Orthogonal())` to `LSTM(weights_init=Orthogonal())` yields a badly initialized network. That's super dangerous.

I propose that regardless of what initialization for 1-D vector we converge on, it should be explicitly activated by a switch, forcing the user to read the docs after performing a switch like I described above. After that I do not care too much what exactly we choose for those degenerate cases, but maybe if there are debates we should not do anything.